bionty/base/_ontology_url.py

```diff
@@ -0,0 +1,168 @@
+from functools import lru_cache
+
+import bioregistry
+import requests
+
+
+class OntologyURLError(Exception):
+    """Base exception for ontology resolver errors."""
+
+    pass
+
+
+class OntologyNotFoundError(OntologyURLError):
+    """Raised when an ontology cannot be found."""
+
+    pass
+
+
+class OntologyVersionNotFoundError(OntologyURLError):
+    """Raised when a specific version of an ontology cannot be found."""
+
+    pass
+
+
+@lru_cache(maxsize=128)
+def get_ontology_url(prefix: str, version: str | None = None) -> tuple[str, str]:
+    """Get a versioned download URL for an ontology based on its CURIE prefix.
+
+    Args:
+        prefix: The CURIE prefix (e.g., 'GO', 'MONDO', 'HP')
+        version: Optional version string (e.g., '2023-01-01')
+                If None, the latest version will be determined.
+
+    Returns:
+        Tuple of (download_url, version_string)
+
+    Raises:
+        OntologyNotFoundError: If the ontology cannot be found
+        OntologyVersionNotFoundError: If no versioned URL can be found
+    """
+    if not prefix:
+        raise ValueError("please provide a prefix")
+
+    # Normalize the prefix
```

Comment:
```suggestion
```

Self documenting code here. Instead, we could consider documenting what the normalization does or why we're normalizing it.

Reply:
> Bioregistry as a dependency brings with it https://github.com/biopragmatics/bioregistry/blob/main/pyproject.toml#L58. They seem to be maintained so for now it should be okay.

I was also a bit annoyed having to introduce it as a dependency, maybe we can keep it in dev for now. I can make it not required if users only use sources hosted on our s3.